### PR TITLE
fix(TDI-47908): Commit stats in case of Error

### PR DESCRIPTION
**Pull request by dmytro-sylaiev**
**What is the current behavior?** (You can also link to an open issue here)

**What is the new behavior?**

**Please check if the PR fulfills these requirements**

- [ ] The commit message follows Talend standard
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features) ?
- [ ] The code coverage on new code >75%
- [ ] The new code does not introduce new technical issues (sonar / eslint)

**What kind of change does this PR introduce?**

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build / CI related changes
- [ ] Other... Please describe:

**Does this PR introduce a breaking change?**

- [ ] Yes
- [ ] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

**Review by dmytrogrygorenko — APPROVED** [review]
LGTM

File: main/plugins/org.talend.designer.codegen/jet_stub/footer.javajet [hunk 1]
```diff
@@ -1156,9 +1156,15 @@ this.globalResumeTicket = false;//to run others jobs
             if (!componentName.equals("tJobStructureCatcher") && !componentName.equals("tLogCatcher") && !componentName.equals("tFlowMeterCatcher") && !componentName.equals("tAssertCatcher") && !componentName.equals("tStatCatcher") && !componentName.equals("tAsyncIn")) {
 %>
 <%=createCallProcess(rootNode, className, false) %>
-<%    }
-    }
-}// end if(isRunInMultiThread)
+<%
+                if (process.getNodesOfType("tStatCatcher").size() > 0) {
+%>
+<%=statsErrorHandlingAfterMainCall(rootNode, process.getNodesOfType("tStatCatcher"))%>
```
Comment (wwang-talend) [inline main/plugins/org.talend.designer.codegen/jet_stub/footer.javajet:1162]: why not change directly in createCallProcess method of footer_java.skeleton? here seems not consider tprejob cases and so on.

File: main/plugins/org.talend.designer.codegen/jet_stub/footer_java.skeleton [hunk 2]
```diff
@@ -55,7 +55,20 @@ public class CLASS
...
-		 
+
+    public String statsErrorHandlingAfterMainCall(INode rootNode, List<? extends INode> statsNodes) {
+        String catchErrorReturn = "catch (Error error_" + rootNode.getUniqueName() + " ) {\n";
+        catchErrorReturn+="end = System.currentTimeMillis();\n";
+        for (INode statCatcherNode : statsNodes) {
+            catchErrorReturn += statCatcherNode.getUniqueName() + ".addMessage(\"failure\", (end-startTime));\n";
+            catchErrorReturn += "try {\n " + statCatcherNode.getDesignSubjobStartNode().getUniqueName() + "Process(globalMap);\n";
+            catchErrorReturn += "} catch (Exception e_" + statCatcherNode.getUniqueName() + ") {\n";
+            catchErrorReturn += "e_" + statCatcherNode.getUniqueName() + ".printStackTrace();\n}\n";
```
Comment (wwang-talend) [inline main/plugins/org.talend.designer.codegen/jet_stub/footer_java.skeleton:66]: here also mean not throw java error, mean the later code can run, need to be careful, please consider this case:
```
                try {
			errorCode = null;
			tFileInputDelimited_1Process(globalMap);
			if (!"failure".equals(status)) {
				status = "end";
			}
		} catch (TalendException e_tFileInputDelimited_1) {
			globalMap.put("tFileInputDelimited_1_SUBPROCESS_STATE", -1);

			e_tFileInputDelimited_1.printStackTrace();

		} catch(Error error) {
                        //1. do state log
                        //2. print error, not throw, mean not stop when error like outofmemory
                 }
```

so some user will ask this: why action is different with and without use stat&logs:
suppose:
A job(with or without multi threads) : 
   subjobA
   subjobB
   subjobC

no subjobok/subjoberror line link to it, before, outofmemory appear, stop at once. Now outofmemory appear, as just ignore it, then we wait for next outofmemory appear..., that is right? 

Also please consdider trunjob case:
A job
   trunjobA(a child job)==>on_subjob_ok==>trunjobB(a child job)
if trunjobA outofmemory, trunjobB will run after current fix? not sure.

File: main/plugins/org.talend.designer.codegen/jet_stub/footer.javajet (hunk 1 above)
Comment (wwang-talend) [reply main/plugins/org.talend.designer.codegen/jet_stub/footer.javajet:1162]: and also here, in jira, user care log table too, seems not fix for that here, as here ony fix for stat table.

File: main/plugins/org.talend.designer.codegen/jet_stub/footer_java.skeleton (hunk 2 above)
Comment (wwang-talend) [reply main/plugins/org.talend.designer.codegen/jet_stub/footer_java.skeleton:66]: In my private view, we should fix this feature firstly before start the bug for stat&log:

decide how to process/stop error.

**Review by wwang-talend — APPROVED** [review]

File: main/plugins/org.talend.designer.codegen/jet_stub/footer.javajet (hunk 1 above)
Comment (dmytro-sylaiev) [reply main/plugins/org.talend.designer.codegen/jet_stub/footer.javajet:1162]: codegen does stats (begin) after prejob, and if Exception occurs in prejob, no stats is printed at all, I respected the consistency here

File: main/plugins/org.talend.designer.codegen/jet_stub/footer_java.skeleton (hunk 2 above)
Comment (dmytro-sylaiev) [reply main/plugins/org.talend.designer.codegen/jet_stub/footer_java.skeleton:66]: We throw errors further and handle and print only TalendException, as without using stats

**Pull request merged**